src/frame/header.rs

```diff
@@ -0,0 +1,342 @@
+const BD_BLOCK_SIZE_MASK_RSHIFT: u8 = 4;
+
+const BLOCK_UNCOMPRESSED_SIZE_BIT: u32 = 0x80000000;
+
+const LZ4F_MAGIC_NUMBER: u32 = 0x184D2204;
+const LZ4F_SKIPPABLE_MAGIC_RANGE: std::ops::RangeInclusive<u32> = 0x184D2A50..=0x184D2A5F;
+
+pub(crate) const MIN_FRAME_INFO_SIZE: usize = 7;
+pub(crate) const MAX_FRAME_INFO_SIZE: usize = 19;
+pub(crate) const BLOCK_INFO_SIZE: usize = 4;
+
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum BlockSize {
+    /// The default block size.
+    Max64KB = 4,
+    Max256KB = 5,
+    Max1MB = 6,
+    Max4MB = 7,
+}
+
+impl Default for BlockSize {
+    fn default() -> Self {
+        BlockSize::Max64KB
+    }
+}
+
+impl BlockSize {
+    pub fn get_size(&self) -> usize {
+        match self {
+            BlockSize::Max64KB => 64 * 1024,
+            BlockSize::Max256KB => 256 * 1024,
+            BlockSize::Max1MB => 1024 * 1024,
+            BlockSize::Max4MB => 4 * 1024 * 1024,
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum BlockMode {
+    /// Every block is compressed independently. The default.
+    Independent,
+    /// Blocks can reference data from previous blocks.
+    ///
+    /// Effective when the stream contains small blocks.
+    Linked,
+}
+
+impl Default for BlockMode {
+    fn default() -> Self {
+        BlockMode::Independent
+    }
+}
+
+/*
+From: https://github.com/lz4/lz4/blob/dev/doc/lz4_Frame_format.md
+
+General Structure of LZ4 Frame format
+-------------------------------------
+
+| MagicNb | F. Descriptor | Block | (...) | EndMark | C. Checksum |
+|:-------:|:-------------:| ----- | ----- | ------- | ----------- |
+| 4 bytes |  3-15 bytes   |       |       | 4 bytes | 0-4 bytes   |
+
+Frame Descriptor
+----------------
+
+| FLG     | BD      | (Content Size) | (Dictionary ID) | HC      |
+| ------- | ------- |:--------------:|:---------------:| ------- |
+| 1 byte  | 1 byte  |  0 - 8 bytes   |   0 - 4 bytes   | 1 byte  |
+
+__FLG byte__
+
+|  BitNb  |  7-6  |   5   |    4     |  3   |    2     |    1     |   0  |
+| ------- |-------|-------|----------|------|----------|----------|------|
+|FieldName|Version|B.Indep|B.Checksum|C.Size|C.Checksum|*Reserved*|DictID|
+
+__BD byte__
+
+|  BitNb  |     7    |     6-5-4     |  3-2-1-0 |
+| ------- | -------- | ------------- | -------- |
+|FieldName|*Reserved*| Block MaxSize |*Reserved*|
+
+Data Blocks
+-----------
+
+| Block Size |  data  | (Block Checksum) |
+|:----------:| ------ |:----------------:|
+|  4 bytes   |        |   0 - 4 bytes    |
+
+*/
+#[derive(Debug, Clone)]
+pub struct FrameInfo {
+    /// If set, includes the total uncompressed size of data in the frame.
+    pub content_size: Option<u64>,
+    /// The identifier for the dictionary that must be used to correctly decode data.
+    /// The compressor and the decompressor must use exactly the same dictionary.
+    ///
+    /// Note that this is currently unsupported and if set will cause de (De)Compressor to error.
+    pub dict_id: Option<u32>,
```

Comment:
I think this should not be pub, as long as it is not supported

Reply:
That's a good idea. Then the FrameEncoder construction can be infallible.